File: forgetask-frontend/app/services/kpiService.ts
/**
 * KPI API Service
 * Handles all KPI calculation and retrieval from the backend
 * This service will call the KPI endpoints for performance metrics
 */
import type { Task } from "@/app/types/task";
import { getApiBaseUrl } from "@/app/services/apiBaseUrl";

export interface KPIMetrics {
  totalTasks: number;
  completedTasks: number;
  backlogCount: number;
  readyCount: number;
  inProgressCount: number;
  reviewCount: number;
  doneCount: number;

  totalEstimatedHours: number;
  totalRealHours: number;
  completedEstimatedHours: number;
  timeVariance: number;
  progressPercentage: number;

  isBacklogOverloaded: boolean;
  isReadyOverloaded: boolean;
  isInProgressOverloaded: boolean;
  isReviewOverloaded: boolean;
  isDoneOverloaded: boolean;
}

export interface KPICalculationRequest {
  tasks: Task[];
  expectedTaskCounts?: Record<string, number>;
}

export interface TaskDistribution {
  [key: string]: number;
}

export interface TimeMetricsSummary {
  totalEstimatedHours: number;
  totalRealHours: number;
  variance: number;
}

export interface RealHoursByUser {
  username: string;
  displayName: string;
  realTotalHours: number;
  doneTasks: number;
}

export interface RealHoursTaskDetail {
  taskId: string;
  title: string;
  realTime: number;
}

export interface RealHoursBySprintUser {
  sprintId: number;
  sprintNumber: number;
  sprintTitle: string;
  username: string;
  displayName: string;
  realTotalHours: number;
  doneTasks: number;
}

export interface ProjectKpisSummary {
  // KPI 1 — TotalTasksKpi
  totalTasks: number;
  tasksBacklog: number;
  tasksReady: number;
  tasksInProgress: number;
  tasksReview: number;
  tasksDone: number;

  // KPI 2 — TotalHoursKpi
  realHours: number;
  estimatedHours: number;

  // KPI 3 — AvgTasksKpi
  totalDevs: number;
  avgTasksPerDev: number;
  sprintTasks?: number;
  sprintDevs?: number;

  // KPI 4 — AvgHoursDevKpi
  avgHoursPerDev: number;
  expectedHoursPerDev: number;
  sprintRealHours?: number;
  sprintEstimatedHours?: number;
}

const API_BASE_URL = getApiBaseUrl();

class KPIService {
  /**
   * Get real worked hours grouped by user.
   * Optionally filtered by sprintId.
   */
  async getRealHoursByUser(sprintId?: number): Promise<RealHoursByUser[]> {
    try {
      const query = sprintId !== undefined ? `?sprintId=${sprintId}` : "";
      const response = await fetch(`${API_BASE_URL}/api/kpi/real-hours-by-user${query}`, {
        method: "GET",
        cache: "no-store",
      });

      if (!response.ok) {
        throw new Error(`Failed to fetch real hours by user: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching real hours by user:", error);
      throw error;
    }
  }

  /**
   * Get done tasks for one user in KPI drill-down.
   * Optionally filtered by sprintId.
   */
  async getRealHoursTasksByUser(
    username: string,
    sprintId?: number
  ): Promise<RealHoursTaskDetail[]> {
    try {
      const params = new URLSearchParams();
      params.set("username", username);
      if (sprintId !== undefined) {
        params.set("sprintId", String(sprintId));
      }

      const response = await fetch(
        `${API_BASE_URL}/api/kpi/real-hours-by-user/tasks?${params.toString()}`,
        {
          method: "GET",
          cache: "no-store",
        }
      );

      if (!response.ok) {
        throw new Error(`Failed to fetch real hours tasks by user: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching real hours tasks by user:", error);
      throw error;
    }
  }

  /**
   * Get general KPI rows where topic is sprint and group is user.
   * Optionally filtered by one sprint.
   */
  async getRealHoursBySprintUser(sprintId?: number): Promise<RealHoursBySprintUser[]> {
    try {
      const params = new URLSearchParams();
      if (sprintId !== undefined) {
        params.set("sprintId", String(sprintId));
      }

      const response = await fetch(
        `${API_BASE_URL}/api/kpi/real-hours-by-sprint-user?${params.toString()}`,
        {
          method: "GET",
          cache: "no-store",
        }
      );

      if (!response.ok) {
        throw new Error(`Failed to fetch real hours by sprint-user: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching real hours by sprint-user:", error);
      throw error;
    }
  }

  /**
   * Calculate KPI metrics from a list of tasks
   * Includes overload detection based on expected task counts
   */
  async calculateKPIs(
    tasks: Task[],
    expectedTaskCounts?: Record<string, number>
  ): Promise<KPIMetrics> {
    try {
      const request: KPICalculationRequest = {
        tasks,
        expectedTaskCounts,
      };

      const response = await fetch(`${API_BASE_URL}/api/kpi/calculate`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(request),
      });

      if (!response.ok) {
        throw new Error(`Failed to calculate KPIs: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      console.error("Error calculating KPIs:", error);
      throw error;
    }
  }

  /**
   * Calculate KPI metrics without overload detection
   * Simplified version for basic KPI calculations
   */
  async calculateKPIsSimple(tasks: Task[]): Promise<KPIMetrics> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/kpi/calculate-simple`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(tasks),
      });

      if (!response.ok) {
        throw new Error(
          `Failed to calculate KPIs: ${response.statusText}`
        );
      }

      return await response.json();
    } catch (error) {
      console.error("Error calculating KPIs:", error);
      throw error;
    }
  }

  /**
   * Get task distribution by status
   */
  async getTaskDistribution(tasks: Task[]): Promise<TaskDistribution> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/kpi/distribution`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(tasks),
      });

      if (!response.ok) {
        throw new Error(
          `Failed to get task distribution: ${response.statusText}`
        );
      }

      return await response.json();
    } catch (error) {
      console.error("Error getting task distribution:", error);
      throw error;
    }
  }

  /**
   * Get time metrics summary
   */
  async getTimeSummary(tasks: Task[]): Promise<TimeMetricsSummary> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/kpi/time-summary`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify(tasks),
      });

      if (!response.ok) {
        throw new Error(
          `Failed to get time summary: ${response.statusText}`
        );
      }

      return await response.json();
    } catch (error) {
      console.error("Error getting time summary:", error);
      throw error;
    }
  }

  /**
   * Check KPI service health
   */
  async checkHealth(): Promise<boolean> {
    try {
      const response = await fetch(`${API_BASE_URL}/api/kpi/health`, {
        method: "GET",
      });

      return response.ok;
    } catch (error) {
      console.error("Error checking KPI service health:", error);
      return false;
    }
  }

  /**
   * Obtiene los 4 KPIs del dashboard en una sola llamada.
   * GET /api/kpi/project/{projectId}/summary
   */
  async getProjectKpisSummary(projectId: number): Promise<ProjectKpisSummary> {
    try {
      const response = await fetch(
        `${API_BASE_URL}/api/kpi/project/${projectId}/summary`,
        {
          method: "GET",
          cache: "no-store", // siempre datos frescos, sin caché
        }
      );

      if (!response.ok) {
        throw new Error(
          `Failed to fetch project KPIs summary: ${response.statusText}`
        );
      }

      return await response.json();
    } catch (error) {
      console.error("Error fetching project KPIs summary:", error);
      throw error;
    }
  }
}

const kpiService = new KPIService();

export default kpiService;

File: forgetask/src/main/java/com/cloudforge/api/forgetask/service/KPIService.java
package com.cloudforge.api.forgetask.service;

import com.cloudforge.api.forgetask.dto.KPIMetrics;
import com.cloudforge.api.forgetask.dto.ProjectKpisSummaryDTO;
import com.cloudforge.api.forgetask.dto.RealHoursByUserDTO;
import com.cloudforge.api.forgetask.dto.RealHoursBySprintUserDTO;
import com.cloudforge.api.forgetask.dto.RealHoursTaskDetailDTO;
import com.cloudforge.api.forgetask.dto.TaskDTO;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Service for calculating KPI metrics from tasks
 * This service is designed to work with any data source (hardcoded, database, etc.)
 * All calculations are based on task status, estimated hours, and real hours
 */
@Service
public class KPIService {

    private final JdbcTemplate jdbcTemplate;

    public KPIService(JdbcTemplate jdbcTemplate) {
        this.jdbcTemplate = jdbcTemplate;
    }

        /**
         * Aggregate users that have tasks and compute done metrics by username.
         * If sprintId is provided, results are scoped to that sprint only.
         */
    public List<RealHoursByUserDTO> getRealHoursByUser(Integer sprintId) {
        String baseSql = """
                SELECT NVL(ua.USERNAME, 'Sin asignar') AS USERNAME,
                  NVL(SUM(CASE WHEN LOWER(NVL(ts.STATE, '')) = 'done' THEN NVL(t.REAL_TIME, 0) ELSE 0 END), 0) AS REAL_TOTAL_HOURS,
                  NVL(SUM(CASE WHEN LOWER(NVL(ts.STATE, '')) = 'done' THEN 1 ELSE 0 END), 0) AS DONE_TASKS
                FROM TASK t
                LEFT JOIN USER_ACCOUNT ua ON ua.ID_USER = t.ID_USER AND ua.ID_PROJECT = t.ID_PROJECT
                LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
              WHERE 1 = 1
                """;

        String orderAndGroup = """
                GROUP BY NVL(ua.USERNAME, 'Sin asignar')
                ORDER BY REAL_TOTAL_HOURS DESC, USERNAME ASC
                """;

        String sql;
        Object[] params;
        if (sprintId != null) {
            sql = baseSql + " AND t.ID_SPRINT = ? " + orderAndGroup;
            params = new Object[] { sprintId };
        } else {
            sql = baseSql + orderAndGroup;
            params = new Object[] {};
        }

        return jdbcTemplate.query(
                sql,
                (rs, rowNum) -> new RealHoursByUserDTO(
                        rs.getString("USERNAME"),
                        rs.getDouble("REAL_TOTAL_HOURS"),
                        rs.getInt("DONE_TASKS")
                ),
                params
        );
    }

    /**
     * Return done hours and done tasks grouped by sprint and user.
     * If sprintId is provided, rows are limited to that sprint.
     */
    public List<RealHoursBySprintUserDTO> getRealHoursBySprintUser(Integer sprintId) {
        String baseSql = """
                SELECT s.ID_SPRINT,
                       COALESCE(
                           CASE
                               WHEN REGEXP_LIKE(s.TITLE, 'Sprint\\s*#\\s*[0-9]+', 'i')
                                   THEN TO_NUMBER(REGEXP_SUBSTR(s.TITLE, 'Sprint\\s*#\\s*([0-9]+)', 1, 1, NULL, 1))
                               ELSE NULL
                           END,
                           s.ID_SPRINT
                       ) AS SPRINT_NUMBER,
                       s.TITLE AS SPRINT_TITLE,
                       NVL(ua.USERNAME, 'Sin asignar') AS USERNAME,
                       NVL(SUM(CASE WHEN LOWER(NVL(ts.STATE, '')) = 'done' THEN NVL(t.REAL_TIME, 0) ELSE 0 END), 0) AS REAL_TOTAL_HOURS,
                       NVL(SUM(CASE WHEN LOWER(NVL(ts.STATE, '')) = 'done' THEN 1 ELSE 0 END), 0) AS DONE_TASKS
                FROM TASK t
                LEFT JOIN USER_ACCOUNT ua ON ua.ID_USER = t.ID_USER AND ua.ID_PROJECT = t.ID_PROJECT
                LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
                LEFT JOIN SPRINT s ON s.ID_SPRINT = t.ID_SPRINT
                WHERE t.ID_SPRINT IS NOT NULL
                """;

        String sprintFilter = sprintId != null ? " AND t.ID_SPRINT = ? " : "";
        String groupAndOrder = " GROUP BY s.ID_SPRINT, s.TITLE, NVL(ua.USERNAME, 'Sin asignar') ORDER BY SPRINT_NUMBER, s.ID_SPRINT, USERNAME ";

        String sql = baseSql + sprintFilter + groupAndOrder;

        Object[] params = sprintId != null ? new Object[] { sprintId } : new Object[] {};

        return jdbcTemplate.query(
                sql,
                (rs, rowNum) -> new RealHoursBySprintUserDTO(
                        rs.getInt("ID_SPRINT"),
                        rs.getInt("SPRINT_NUMBER"),
                        rs.getString("SPRINT_TITLE"),
                        rs.getString("USERNAME"),
                        rs.getDouble("REAL_TOTAL_HOURS"),
                        rs.getInt("DONE_TASKS")
                ),
                params
        );
    }

    /**
     * Return done tasks (title + real time) for one user.
     * If sprintId is provided, rows are limited to that sprint.
     */
    public List<RealHoursTaskDetailDTO> getRealHoursTasksByUser(String username, Integer sprintId) {
        String baseSql = """
                SELECT TO_CHAR(t.ID_TASK) AS TASK_ID,
                       t.TITLE,
                       NVL(t.REAL_TIME, 0) AS REAL_TIME
                FROM TASK t
                LEFT JOIN USER_ACCOUNT ua ON ua.ID_USER = t.ID_USER AND ua.ID_PROJECT = t.ID_PROJECT
                LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
                WHERE LOWER(ts.STATE) = 'done'
                """;

        String userFilter;
        if ("Sin asignar".equalsIgnoreCase(username)) {
            userFilter = " AND ua.USERNAME IS NULL ";
        } else {
            userFilter = " AND ua.USERNAME = ? ";
        }

        String sprintFilter = sprintId != null ? " AND t.ID_SPRINT = ? " : "";
        String orderBy = " ORDER BY NVL(t.REAL_TIME, 0) DESC, t.ID_TASK DESC ";

        String sql = baseSql + userFilter + sprintFilter + orderBy;

        Object[] params;
        if ("Sin asignar".equalsIgnoreCase(username)) {
            if (sprintId != null) {
                params = new Object[] { sprintId };
            } else {
                params = new Object[] {};
            }
        } else {
            if (sprintId != null) {
                params = new Object[] { username, sprintId };
            } else {
                params = new Object[] { username };
            }
        }

        return jdbcTemplate.query(
                sql,
                (rs, rowNum) -> new RealHoursTaskDetailDTO(
                        rs.getString("TASK_ID"),
                        rs.getString("TITLE"),
                        rs.getDouble("REAL_TIME")
                ),
                params
        );
    }

    /**
     * Calculate KPI metrics from a list of tasks
     * @param tasks List of tasks to calculate KPIs from
     * @param expectedTaskCounts Map of status to expected task count (optional for overload detection)
     * @return KPIMetrics object with all calculated KPIs
     */
    public KPIMetrics calculateKPIs(List<TaskDTO> tasks, Map<String, Integer> expectedTaskCounts) {
        KPIMetrics metrics = new KPIMetrics();

        if (tasks == null || tasks.isEmpty()) {
            return metrics;
        }

        // Initialize counts and sums
        int backlogCount = 0;
        int readyCount = 0;
        int inProgressCount = 0;
        int reviewCount = 0;
        int doneCount = 0;
        double totalEstimatedHours = 0.0;
        double totalRealHours = 0.0;
        double completedEstimatedHours = 0.0;

        // Process each task
        for (TaskDTO task : tasks) {
            String status = task.getStatus();
            double estimatedTime = task.getEstimatedTime();
            double realTime = task.getRealTime();

            // Count by status
            if ("backlog".equalsIgnoreCase(status)) {
                backlogCount++;
            } else if ("ready".equalsIgnoreCase(status)) {
                readyCount++;
            } else if ("in-progress".equalsIgnoreCase(status)) {
                inProgressCount++;
            } else if ("review".equalsIgnoreCase(status)) {
                reviewCount++;
            } else if ("done".equalsIgnoreCase(status)) {
                doneCount++;
                completedEstimatedHours += estimatedTime;
            }

            // Sum hours
            totalEstimatedHours += estimatedTime;
            totalRealHours += realTime;
        }

        // Set basic counts
        metrics.setTotalTasks(tasks.size());
        metrics.setCompletedTasks(doneCount);
        metrics.setBacklogCount(backlogCount);
        metrics.setReadyCount(readyCount);
        metrics.setInProgressCount(inProgressCount);
        metrics.setReviewCount(reviewCount);
        metrics.setDoneCount(doneCount);

        // Set time metrics
        metrics.setTotalEstimatedHours(totalEstimatedHours);
        metrics.setTotalRealHours(totalRealHours);
        metrics.setCompletedEstimatedHours(completedEstimatedHours);
        metrics.setTimeVariance(totalRealHours - totalEstimatedHours);

        // Calculate progress percentage
        int progressPercentage = totalEstimatedHours > 0
                ? Math.round((float) (completedEstimatedHours / totalEstimatedHours) * 100)
                : 0;
        metrics.setProgressPercentage(progressPercentage);

        // Detect overloaded columns based on expected counts
        if (expectedTaskCounts != null && !expectedTaskCounts.isEmpty()) {
            metrics.setBacklogOverloaded(backlogCount > expectedTaskCounts.getOrDefault("backlog", Integer.MAX_VALUE));
            metrics.setReadyOverloaded(readyCount > expectedTaskCounts.getOrDefault("ready", Integer.MAX_VALUE));
            metrics.setInProgressOverloaded(inProgressCount > expectedTaskCounts.getOrDefault("in-progress", Integer.MAX_VALUE));
            metrics.setReviewOverloaded(reviewCount > expectedTaskCounts.getOrDefault("review", Integer.MAX_VALUE));
            metrics.setDoneOverloaded(doneCount > expectedTaskCounts.getOrDefault("done", Integer.MAX_VALUE));
        }

        return metrics;
    }

    /**
     * Simplified version - calculate KPIs without overload detection
     * @param tasks List of tasks to calculate KPIs from
     * @return KPIMetrics object with all calculated KPIs
     */
    public KPIMetrics calculateKPIs(List<TaskDTO> tasks) {
        return calculateKPIs(tasks, new HashMap<>());
    }

    /**
     * Calculate task distribution by status
     * @param tasks List of tasks
     * @return Map with status as key and count as value
     */
    public Map<String, Integer> getTaskDistributionByStatus(List<TaskDTO> tasks) {
        Map<String, Integer> distribution = new HashMap<>();
        distribution.put("backlog", 0);
        distribution.put("ready", 0);
        distribution.put("in-progress", 0);
        distribution.put("review", 0);
        distribution.put("done", 0);

        if (tasks == null || tasks.isEmpty()) {
            return distribution;
        }

        for (TaskDTO task : tasks) {
            String status = task.getStatus();
            distribution.put(status, distribution.getOrDefault(status, 0) + 1);
        }

        return distribution;
    }

    /**
     * Calculate time-based metrics summary
     * @param tasks List of tasks
     * @return Map with time metrics
     */
    public Map<String, Double> getTimeMetricsSummary(List<TaskDTO> tasks) {
        Map<String, Double> summary = new HashMap<>();
        summary.put("totalEstimatedHours", 0.0);
        summary.put("totalRealHours", 0.0);
        summary.put("variance", 0.0);

        if (tasks == null || tasks.isEmpty()) {
            return summary;
        }

        double totalEstimated = 0.0;
        double totalReal = 0.0;

        for (TaskDTO task : tasks) {
            totalEstimated += task.getEstimatedTime();
            totalReal += task.getRealTime();
        }

        summary.put("totalEstimatedHours", totalEstimated);
        summary.put("totalRealHours", totalReal);
        summary.put("variance", totalReal - totalEstimated);

        return summary;
    }

    /**
     * Calcula todos los KPIs del dashboard para un proyecto dado.
     * Consulta directamente la BD sin necesitar que el frontend envíe datos.
     */
    public ProjectKpisSummaryDTO getProjectKpisSummary(Integer projectId) {

        // ── KPI 1: Conteo de tasks por estado ──
        // TASK_STATE tiene los estados: backlog, ready, in_progress, review, done
        String taskCountsSql = """
            SELECT
                COUNT(t.ID_TASK) AS TOTAL_TASKS,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'backlog'     THEN 1 ELSE 0 END), 0) AS BACKLOG,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'ready'       THEN 1 ELSE 0 END), 0) AS READY,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'in_progress' THEN 1 ELSE 0 END), 0) AS IN_PROGRESS,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'review'      THEN 1 ELSE 0 END), 0) AS REVIEW,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'done'        THEN 1 ELSE 0 END), 0) AS DONE
            FROM TASK t
            LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
            WHERE t.ID_PROJECT = ?
            """;

        Map<String, Object> taskCounts = jdbcTemplate.queryForMap(taskCountsSql, projectId);

        int totalTasks    = ((Number) taskCounts.get("TOTAL_TASKS")).intValue();
        int tasksBacklog  = ((Number) taskCounts.get("BACKLOG")).intValue();
        int tasksReady    = ((Number) taskCounts.get("READY")).intValue();
        int tasksInProgress = ((Number) taskCounts.get("IN_PROGRESS")).intValue();
        int tasksReview   = ((Number) taskCounts.get("REVIEW")).intValue();
        int tasksDone     = ((Number) taskCounts.get("DONE")).intValue();

        // ── KPI 2: Horas reales vs estimadas ──
        // Leídas directamente de PROJECT (las columnas ya existen en tu tabla)
        String hoursSql = """
            SELECT
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'done' THEN t.REAL_TIME ELSE 0 END), 0) AS REAL_TIME,
                NVL(SUM(t.ESTIMATED_TIME), 0) AS ESTIMATED_TIME
            FROM TASK t
            LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
            WHERE t.ID_PROJECT = ?
            """;

        Map<String, Object> hours = jdbcTemplate.queryForMap(hoursSql, projectId);
        double realHours      = ((Number) hours.get("REAL_TIME")).doubleValue();
        double estimatedHours = ((Number) hours.get("ESTIMATED_TIME")).doubleValue();

        // ── KPI 3 y 4: Número de developers del proyecto ──
        // Solo contamos usuarios con ROLE = 'developer', no managers
        String devCountSql = """
            SELECT COUNT(DISTINCT ua.ID_USER) AS TOTAL_DEVS
            FROM USER_ACCOUNT ua
            JOIN USER_ROLE ur ON ur.ID_USER = ua.ID_USER
            WHERE ua.ID_PROJECT = ?
            AND LOWER(ur.ROLE) = 'developer'
            """;

        Integer totalDevsResult = jdbcTemplate.queryForObject(devCountSql, Integer.class, projectId);
        int totalDevs = (totalDevsResult != null) ? totalDevsResult : 0;

        // ── KPI 3: Promedio tasks por developer ──
        double avgTasksPerDev = totalDevs > 0
            ? Math.round((double) totalTasks / totalDevs * 10.0) / 10.0
            : 0;
        
        String sprintTasksSql = """
            SELECT COUNT(t.ID_TASK) AS SPRINT_TOTAL_TASKS
            FROM TASK t
            WHERE t.ID_PROJECT = ?
            AND t.ID_SPRINT = NVL((
                SELECT ID_SPRINT FROM (
                    SELECT s.ID_SPRINT
                    FROM SPRINT s
                    WHERE s.ID_PROJECT = ?
                        AND s.START_DATE <= SYSDATE
                        AND s.END_DATE   >= SYSDATE
                ) WHERE ROWNUM = 1
            ), -1)
            """;
        
        Integer sprintTasksResult = jdbcTemplate.queryForObject(
            sprintTasksSql, Integer.class, projectId, projectId
        );
        int sprintTasks = sprintTasksResult != null ? sprintTasksResult : 0;

        // ── KPI 4: Promedio horas por developer ──
        double avgHoursPerDev = totalDevs > 0
            ? Math.round(realHours / totalDevs * 10.0) / 10.0
            : 0;

        double expectedHoursPerDev = totalDevs > 0
            ? Math.round(estimatedHours / totalDevs * 10.0) / 10.0
            : 0;

        String sprintHoursSql = """
            SELECT
                NVL(SUM(t.ESTIMATED_TIME), 0) AS SPRINT_ESTIMATED_HOURS,
                NVL(SUM(CASE WHEN LOWER(ts.STATE) = 'done' THEN t.REAL_TIME ELSE 0 END), 0) AS SPRINT_REAL_HOURS
            FROM TASK t
            LEFT JOIN TASK_STATE ts ON ts.ID_TASK = t.ID_TASK
            WHERE t.ID_PROJECT = ?
            AND t.ID_SPRINT = NVL((
                SELECT ID_SPRINT FROM (
                    SELECT s.ID_SPRINT
                    FROM SPRINT s
                    WHERE s.ID_PROJECT = ?
                        AND s.START_DATE <= SYSDATE
                        AND s.END_DATE   >= SYSDATE
                ) WHERE ROWNUM = 1
            ), -1)
            """;

        Map<String, Object> sprintHours = jdbcTemplate.queryForMap(
            sprintHoursSql, projectId, projectId
        );
        double sprintEstimatedHours = ((Number) sprintHours.get("SPRINT_ESTIMATED_HOURS")).doubleValue();
        double sprintRealHours      = ((Number) sprintHours.get("SPRINT_REAL_HOURS")).doubleValue();

        return new ProjectKpisSummaryDTO(
            totalTasks, tasksBacklog, tasksReady, tasksInProgress, tasksReview, tasksDone,
            realHours, estimatedHours,
            totalDevs,
            avgTasksPerDev,
            sprintTasks,
            totalDevs,
            avgHoursPerDev, expectedHoursPerDev, sprintRealHours, sprintEstimatedHours
        );
    }
}